src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -466,7 +524,44 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
             if (NameResolutionPal.SupportsGetAddrInfoAsync && ipAddress is null)
             {
                 ValidateHostName(hostName);
-                return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+
+                if (NameResolutionTelemetry.IsEnabled)
+                {
+                    ValueStopwatch stopwatch = NameResolutionTelemetry.Log.ResolutionStart(hostName);
+
+                    Task coreTask;
+                    try
+                    {
+                        coreTask = NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+                    }
+                    catch when (LogFailure(hostName, stopwatch))
+                    {
+                        Debug.Fail("LogFailure should return false");
+                        throw;
+                    }
+
+                    coreTask.ContinueWith(
+                        (task, state) =>
+                        {
+                            var tuple = (Tuple<string, ValueStopwatch>)state!;
+                            NameResolutionTelemetry.Log.AfterResolution(
+                                hostNameOrAddress: tuple.Item1,
+                                stopwatch: tuple.Item2,
+                                successful: task.IsCompletedSuccessfully);
+                        },
+                        state: Tuple.Create(hostName, stopwatch),
+                        cancellationToken: default,
+                        TaskContinuationOptions.ExecuteSynchronously,
+                        TaskScheduler.Default);
+
+                    // coreTask is not actually a base Task, but Task<IPHostEntry> / Task<IPAddress[]>
+                    // We have to return it and not the continuation
+                    return coreTask;
+                }
```

Comment:
I'd rather this be a separate async method, e.g.
```C#
return NameResolutionTelemetry.IsEnabled ?
    WithLogging(hostName, justReturnParsedIp, throwOnIPAny, justAddresses) :
    NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);

static async Task WithLogging(string hostName, bool ustAddresses)
{
    ValueStopwatch stopwatch = NameResolutionTelemetry.Log.ResolutionStart(hostName);
    bool successful = true;
    try
    {
        await NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
    }
    catch
    {
        LogFailure(hostName, stopwatch);
        successful = false;
    }

    NameResolutionTelemetry.Log.AfterResolution(hostName, stopwatch, successful);
}
```

Reply:
Ah, I just saw your comment at the end. Ok.

Reply:
But `NameResolutionPal.GetAddrInfoAsync` returns a `Task` object that is actually `Task<IPHostEntry>` / `Task<IPAddress[]>`, so we have to return that task itself

Reply:
Unresolving this to make sure the comment I added is visible, resolve again any time